tests/js/server/aql/aql-traversal-paths-optimization.js

```diff
@@ -0,0 +1,216 @@
+/// @brief tests for traversal optimization
+///
+/// DISCLAIMER
+///
+/// Copyright 2010-2014 triagens GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is triAGENS GmbH, Cologne, Germany
+///
+/// @author Jan Steemann
+/// @author Copyright 2020, triAGENS GmbH, Cologne, Germany
+////////////////////////////////////////////////////////////////////////////////
+
+const jsunity = require("jsunity");
+const db = require("@arangodb").db;
+const { deriveTestSuite } = require('@arangodb/test-helper');
+const isCoordinator = require('@arangodb/cluster').isCoordinator();
+const isEnterprise = require("internal").isEnterprise();
+  
+const vn = "UnitTestsVertex";
+const en = "UnitTestsEdges";
+const gn = "UnitTestsGraph";
+
+function BaseTestConfig() {
+  return {
+
+    testProducePathsOutputs: function () {
+      [
+        [ "FOR v, e, p IN 1..1 OUTBOUND '" + vn + "/test0' GRAPH '" + gn + "' RETURN p", [ true, true, false ]], 
+        [ "FOR v, e, p IN 1..1 OUTBOUND '" + vn + "/test0' GRAPH '" + gn + "' RETURN p[0]", [ true, true, false ]],
+        [ "FOR v, e, p IN 1..1 OUTBOUND '" + vn + "/test0' GRAPH '" + gn + "' RETURN LENGTH(p)", [ true, true, false ]],
+        [ "FOR v, e, p IN 1..1 OUTBOUND '" + vn + "/test0' GRAPH '" + gn + "' RETURN p.edges", [ false, true, false ]],
+        [ "FOR v, e, p IN 1..1 OUTBOUND '" + vn + "/test0' GRAPH '" + gn + "' RETURN p.vertices", [ true, false, false ]],
+        [ "FOR v, e, p IN 1..1 OUTBOUND '" + vn + "/test0' GRAPH '" + gn + "' RETURN [p.edges, p.vertices]", [ true, true, false ]],
+        [ "FOR v, e, p IN 1..1 OUTBOUND '" + vn + "/test0' GRAPH '" + gn + "' RETURN [p.edges, p]", [ true, true, false ]],
+        [ "FOR v, e, p IN 1..1 OUTBOUND '" + vn + "/test0' GRAPH '" + gn + "' RETURN [p.vertices, p]", [ true, true, false ]],
+        [ "FOR v, e, p IN 1..1 OUTBOUND '" + vn + "/test0' GRAPH '" + gn + "' RETURN [p.edges, p.vertices, p]", [ true, true, false ]],
+        [ "FOR v, e, p IN 1..1 OUTBOUND '" + vn + "/test0' GRAPH '" + gn + "' RETURN p.testi", [ true, false, false ]],
+        [ "FOR v, e, p IN 1..1 OUTBOUND '" + vn + "/test0' GRAPH '" + gn + "' COLLECT pl = LENGTH(p) RETURN pl", [ true, true, false ]],
+        [ "FOR v, e, p IN 1..1 OUTBOUND '" + vn + "/test0' GRAPH '" + gn + "' COLLECT key = v._key INTO g RETURN { key, other: g[*].p.edges }", [ true, true, false ]],
```

Comment:
Nice this case is good.
Would love to move it to "false, true, false" at one Point. But I agree that is super hard to figure out.